File: sections/70_bounded_min_cost_flows/problems/d_min_cost_bounded_circulation/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using int64 = long long;
const int64 INF = (1LL << 62);

struct MinCostFlow {
    struct Edge {
        int to, reverse, capacity;
        int64 cost;
    };
    vector<vector<Edge>> graph;

    explicit MinCostFlow(int n) : graph(n) {}

    void add_edge(int u, int v, int capacity, int64 cost) {
        int ui = graph[u].size();
        int vi = graph[v].size();
        graph[u].push_back({v, vi, capacity, cost});
        graph[v].push_back({u, ui, 0, -cost});
    }

    pair<int, int64> flow(int source, int sink, int required) {
        int n = graph.size();
        int sent = 0;
        int64 total_cost = 0;
        vector<int64> potential(n), distance(n);
        vector<int> previous_vertex(n), previous_edge(n);

        while (sent < required) {
            fill(distance.begin(), distance.end(), INF);
            distance[source] = 0;
            priority_queue<pair<int64, int>, vector<pair<int64, int>>,
                           greater<pair<int64, int>>> queue;
            queue.push({0, source});

            while (!queue.empty()) {
                auto [current_distance, u] = queue.top();
                queue.pop();
                if (current_distance != distance[u]) continue;
                for (int index = 0; index < (int)graph[u].size(); ++index) {
                    const Edge& edge = graph[u][index];
                    if (!edge.capacity) continue;
                    int64 candidate = current_distance + edge.cost
                                    + potential[u] - potential[edge.to];
                    if (candidate < distance[edge.to]) {
                        distance[edge.to] = candidate;
                        previous_vertex[edge.to] = u;
                        previous_edge[edge.to] = index;
                        queue.push({candidate, edge.to});
                    }
                }
            }
            if (distance[sink] == INF) break;
            for (int v = 0; v < n; ++v) {
                if (distance[v] < INF) potential[v] += distance[v];
            }

            int add = required - sent;
            for (int v = sink; v != source; v = previous_vertex[v]) {
                add = min(add, graph[previous_vertex[v]][previous_edge[v]].capacity);
            }
            for (int v = sink; v != source; v = previous_vertex[v]) {
                Edge& edge = graph[previous_vertex[v]][previous_edge[v]];
                total_cost += int64(add) * edge.cost;
                edge.capacity -= add;
                graph[v][edge.reverse].capacity += add;
            }
            sent += add;
        }
        return {sent, total_cost};
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;
    int source = n;
    int sink = n + 1;
    MinCostFlow network(n + 2);
    vector<int> balance(n);
    int64 mandatory_cost = 0;

    for (int edge = 0; edge < m; ++edge) {
        int u, v, low, high, cost;
        cin >> u >> v >> low >> high >> cost;
        --u;
        --v;
        balance[u] -= low;
        balance[v] += low;
        mandatory_cost += int64(low) * cost;
        network.add_edge(u, v, high - low, cost);
    }

    int required = 0;
    for (int v = 0; v < n; ++v) {
        if (balance[v] > 0) {
            network.add_edge(source, v, balance[v], 0);
            required += balance[v];
        } else if (balance[v] < 0) {
            network.add_edge(v, sink, -balance[v], 0);
        }
    }

    auto [sent, extra_cost] = network.flow(source, sink, required);
    if (sent != required) cout << "IMPOSSIBLE\n";
    else cout << mandatory_cost + extra_cost << '\n';
}

File: sections/70_bounded_min_cost_flows/problems/d_min_cost_bounded_circulation/solution.py
import heapq
import sys


class MinCostFlow:
    def __init__(self, n: int):
        self.graph = [[] for _ in range(n)]

    def add_edge(self, u: int, v: int, capacity: int, cost: int) -> None:
        self.graph[u].append([v, capacity, cost, len(self.graph[v])])
        self.graph[v].append([u, 0, -cost, len(self.graph[u]) - 1])

    def flow(self, source: int, sink: int, required: int):
        n = len(self.graph)
        potential = [0] * n
        sent = total_cost = 0
        infinity = 10**30

        while sent < required:
            distance = [infinity] * n
            previous = [None] * n
            distance[source] = 0
            queue = [(0, source)]
            while queue:
                current_distance, u = heapq.heappop(queue)
                if current_distance != distance[u]:
                    continue
                for index, edge in enumerate(self.graph[u]):
                    v, capacity, cost, _ = edge
                    if not capacity:
                        continue
                    candidate = current_distance + cost + potential[u] - potential[v]
                    if candidate < distance[v]:
                        distance[v] = candidate
                        previous[v] = (u, index)
                        heapq.heappush(queue, (candidate, v))
            if distance[sink] == infinity:
                break
            for v in range(n):
                if distance[v] < infinity:
                    potential[v] += distance[v]

            add = required - sent
            v = sink
            while v != source:
                u, index = previous[v]
                add = min(add, self.graph[u][index][1])
                v = u
            v = sink
            while v != source:
                u, index = previous[v]
                edge = self.graph[u][index]
                total_cost += add * edge[2]
                edge[1] -= add
                self.graph[v][edge[3]][1] += add
                v = u
            sent += add
        return sent, total_cost


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n, m = data[:2]
    source, sink = n, n + 1
    network = MinCostFlow(n + 2)
    balance = [0] * n
    mandatory_cost = 0
    at = 2
    for _ in range(m):
        u, v, low, high, cost = data[at : at + 5]
        at += 5
        u -= 1
        v -= 1
        balance[u] -= low
        balance[v] += low
        mandatory_cost += low * cost
        network.add_edge(u, v, high - low, cost)

    required = 0
    for v, amount in enumerate(balance):
        if amount > 0:
            network.add_edge(source, v, amount, 0)
            required += amount
        elif amount < 0:
            network.add_edge(v, sink, -amount, 0)

    sent, extra_cost = network.flow(source, sink, required)
    print(mandatory_cost + extra_cost if sent == required else "IMPOSSIBLE")


if __name__ == "__main__":
    main()
